File: src/01_types/Delta.jl
# src/types/Delta.jl
"""
Delta - Represents a byte-level state change

Deltas capture minimal change sets for page updates.
Contains only changed bytes with mask + data.
"""
module DeltaTypes

using Serialization
using ..PageTypes: PageID

export Delta, DeltaID, delta_byte_count, compress_delta, merge_deltas,
       dense_data, is_sparse_delta, serialize_delta, deserialize_delta

const DeltaID = UInt64
const SPARSE_THRESHOLD = 0.25

"""
    Delta

Represents a byte-level change to a page.
"""
struct Delta
    id::DeltaID
    page_id::PageID
    epoch::UInt32
    mask::BitVector
    data::Vector{UInt8}
    timestamp::UInt64
    source::Symbol
    
    function Delta(id::DeltaID, page_id::PageID, epoch::UInt32, 
                   mask::AbstractVector{Bool}, data::AbstractVector{UInt8}, source::Symbol)
        @assert length(mask) == length(data) "Delta mask/data length mismatch"
        mask_vec = BitVector(mask)
        payload = _compress_payload(mask_vec, Vector{UInt8}(data))
        return new(
            id,
            page_id,
            epoch,
            mask_vec,
            payload,
            time_ns(),
            source,
        )
    end
    
    function Delta(id::DeltaID, page_id::PageID, epoch::UInt32, 
                   mask::AbstractVector{Bool}, data::AbstractVector{UInt8}, 
                   source::Symbol, timestamp::UInt64)
        @assert length(mask) == length(data) || length(data) == count(identity, mask)
        mask_vec = BitVector(mask)
        payload = Vector{UInt8}(data)
        return new(id, page_id, epoch, mask_vec, payload, timestamp, source)
    end
end

"""
Decide whether to store sparse or dense payloads and build buffer accordingly.
"""
function _compress_payload(mask::BitVector, data::Vector{UInt8})::Vector{UInt8}
    changed = count(identity, mask)
    changed == 0 && return UInt8[]
    ratio = changed / length(mask)
    if ratio <= SPARSE_THRESHOLD
        compressed = Vector{UInt8}(undef, changed)
        idx = 1
        @inbounds for i in eachindex(mask)
            if mask[i]
                compressed[idx] = data[i]
                idx += 1
            end
        end
        return compressed
    else
        return copy(data)
    end
end

"""
    delta_byte_count(delta::Delta) -> Int

Count number of changed bytes in delta.
"""
delta_byte_count(delta::Delta)::Int = count(identity, delta.mask)

"""
    is_sparse_delta(delta::Delta) -> Bool

Return true if the delta stores only changed bytes.
"""
is_sparse_delta(delta::Delta)::Bool = length(delta.data) != length(delta.mask)

"""
    dense_data(delta::Delta) -> Vector{UInt8}

Materialize a dense byte vector (length == mask length) for a delta.
"""
function dense_data(delta::Delta)::Vector{UInt8}
    len = length(delta.mask)
    if !is_sparse_delta(delta)
        return copy(delta.data)
    end
    dense = zeros(UInt8, len)
    idx = 1
    @inbounds for i in 1:len
        if delta.mask[i]
            dense[i] = delta.data[idx]
            idx += 1
        end
    end
    return dense
end

"""
    compress_delta(delta::Delta) -> Delta

Create a compressed version of delta (sparse representation).
"""
function compress_delta(delta::Delta)::Delta
    dense = dense_data(delta)
    return Delta(
        delta.id,
        delta.page_id,
        delta.epoch,
        delta.mask,
        dense,
        delta.source,
    )
end

"""
    merge_deltas(d1::Delta, d2::Delta) -> Delta

Merge two deltas targeting the same page with d2 taking precedence.
"""
function merge_deltas(d1::Delta, d2::Delta)::Delta
    @assert d1.page_id == d2.page_id "Cannot merge deltas across pages"
    merged_mask = copy(d1.mask)
    merged_data = dense_data(d1)
    d2_dense = dense_data(d2)
    @inbounds for i in eachindex(d2.mask)
        if d2.mask[i]
            merged_mask[i] = true
            merged_data[i] = d2_dense[i]
        end
    end
    new_epoch = max(d1.epoch, d2.epoch)
    return Delta(d2.id, d2.page_id, new_epoch, merged_mask, merged_data, d2.source)
end

"""
    serialize_delta(delta::Delta) -> Vector{UInt8}

Serialize a delta to a compact byte vector.
"""
function serialize_delta(delta::Delta)::Vector{UInt8}
    payload = (
        delta.id,
        delta.page_id,
        delta.epoch,
        Vector{Bool}(delta.mask),
        delta.data,
        length(delta.data) != length(delta.mask),
        delta.timestamp,
        delta.source,
    )
    io = IOBuffer()
    Serialization.serialize(io, payload)
    return take!(io)
end

"""
    deserialize_delta(bytes::Vector{UInt8}) -> Delta

Reconstruct delta from serialized bytes.
"""
function deserialize_delta(bytes::Vector{UInt8})::Delta
    io = IOBuffer(bytes)
    id, page_id, epoch, mask_vec, data, is_sparse, timestamp, source = Serialization.deserialize(io)
    mask = BitVector(mask_vec)
    if !is_sparse && length(data) != length(mask)
        resize!(data, length(mask))
    end
    return Delta(id, page_id, epoch, mask, data, source, timestamp)
end

end # module DeltaTypes

File: src/01_types/delta.rs
use super::epoch::Epoch;
use super::page::{Page, PageError, PageID};
use std::time::{SystemTime, UNIX_EPOCH};

#[repr(transparent)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct DeltaID(pub u64);

#[derive(Debug, Clone)]
pub struct Source(pub String);

#[derive(Debug, Clone)]
pub struct Delta {
    pub delta_id: DeltaID,
    pub page_id: PageID,
    pub epoch: Epoch,
    pub mask: Vec<bool>,
    pub payload: Vec<u8>,
    pub is_sparse: bool,
    pub timestamp: u64,
    pub source: Source,
}

impl Delta {
    pub fn new_dense(
        delta_id: DeltaID,
        page_id: PageID,
        epoch: Epoch,
        data: Vec<u8>,
        mask: Vec<bool>,
        source: Source,
    ) -> Result<Self, DeltaError> {
        if data.len() != mask.len() {
            return Err(DeltaError::SizeMismatch {
                mask_len: mask.len(),
                payload_len: data.len(),
            });
        }

        Ok(Self {
            delta_id,
            page_id,
            epoch,
            mask,
            payload: data,
            is_sparse: false,
            timestamp: now_ns(),
            source,
        })
    }

    pub fn new_sparse(
        delta_id: DeltaID,
        page_id: PageID,
        epoch: Epoch,
        mask: Vec<bool>,
        payload: Vec<u8>,
        source: Source,
    ) -> Result<Self, DeltaError> {
        let changed = mask.iter().filter(|&&m| m).count();
        if changed != payload.len() {
            return Err(DeltaError::SizeMismatch {
                mask_len: mask.len(),
                payload_len: payload.len(),
            });
        }

        Ok(Self {
            delta_id,
            page_id,
            epoch,
            mask,
            payload,
            is_sparse: true,
            timestamp: now_ns(),
            source,
        })
    }

    pub fn merge(&self, other: &Delta) -> Result<Delta, DeltaError> {
        if self.page_id != other.page_id {
            return Err(DeltaError::PageIDMismatch {
                expected: self.page_id,
                found: other.page_id,
            });
        }

        if self.mask.len() != other.mask.len() {
            return Err(DeltaError::MaskSizeMismatch {
                expected: self.mask.len(),
                found: other.mask.len(),
            });
        }

        let mut merged_mask = self.mask.clone();
        let mut merged_payload = self.to_dense();
        let other_dense = other.to_dense();

        for (idx, &flag) in other.mask.iter().enumerate() {
            if flag {
                merged_mask[idx] = true;
                merged_payload[idx] = other_dense[idx];
            }
        }

        Ok(Delta {
            delta_id: other.delta_id,
            page_id: other.page_id,
            epoch: Epoch(other.epoch.0.max(self.epoch.0)),
            mask: merged_mask,
            payload: merged_payload,
            is_sparse: false,
            timestamp: other.timestamp.max(self.timestamp),
            source: other.source.clone(),
        })
    }

    pub fn to_dense(&self) -> Vec<u8> {
        if !self.is_sparse {
            return self.payload.clone();
        }

        let mut dense = vec![0u8; self.mask.len()];
        let mut payload_idx = 0;
        for (idx, &flag) in self.mask.iter().enumerate() {
            if flag {
                dense[idx] = self.payload[payload_idx];
                payload_idx += 1;
            }
        }
        dense
    }

    pub fn apply_to(&self, page: &mut Page) -> Result<(), PageError> {
        page.apply_delta(self)
    }
}

#[derive(Debug, thiserror::Error)]
pub enum DeltaError {
    #[error("Mask/payload size mismatch mask={mask_len} payload={payload_len}")]
    SizeMismatch { mask_len: usize, payload_len: usize },

    #[error("PageID mismatch: expected {expected:?}, found {found:?}")]
    PageIDMismatch { expected: PageID, found: PageID },

    #[error("Mask size mismatch: expected {expected}, found {found}")]
    MaskSizeMismatch { expected: usize, found: usize },
}

fn now_ns() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_nanos() as u64
}
